### ISSUE TITLE
support passing custom arguments to cargo test

### ISSUE THREAD

**Issue by hardfist** [rust-lang/cargo#11339]
### Problem

We want to support custom snapshot feature to our test, 
and we need a way to control whether we in test mode or in update mode.
There are currently three ways to implement to this.
* implement a custom command like cargo insta 
* control it by environment like UPDATE to control whether we are in test mode or in update mode
* disable harnessand convert all test to main function to run test
They are both not the idea way, and we want something like jest or mocha.
```sh
$ jest  // test mode
$ jest --updateSnapshot // update mode
$ mocha // test mode
$ mocha --update // update mode
```
if cargo test could support passing custom arguments to test, we could do it like jest
```sh
$ cargo test // test mode
$ cargo test --update // update mode
```

### Proposed Solution

support passing custom arguments to cargo test, maybe we need a configuration for this feature.

### Notes

someone also met same problem but not got a ideal answer
* https://stackoverflow.com/questions/65791697/passing-custom-command-line-arguments-to-a-rust-test
* https://stackoverflow.com/questions/72804945/passing-arguments-to-cargo-test-with-clap

**Comment by epage** [comment]
There are two workarounds

- env variables
- using the positional argument for libtest: https://github.com/assert-rs/trycmd/blob/main/src/cases.rs#L190

Challenges with improving this at cargo's level
- Most of this actually lives in libtest
- clap doesn't support this yet
- Even if clap supports it, there are weird parse cases (unknown flag followed by a value, is that value a positional argument or meant for the flag)

**Comment by ehuss** [comment]
There are some issues that I would be concerned about with extensible flags:

* I think it would be difficult to be able to convey "custom flags" to tell Cargo to parse. Additionally, I'm not even sure that is a good thing, as I think having a standard experience for which flags are valid is good for consistency and avoiding confusion.
* libtest similarly has the same issue, where it won't know what flags your test is interested in, and which ones it should be parsing.

I understand this is a fairly common use case, so it might be nice to have something to make it easier. I think the `cargo insta` approach isn't too bad, and I'm curious what makes it seem not ideal.  Environment variables I think are a common approach, and I'm also curious what makes them seem not ideal. I can make some guesses, but I'd like to hear more about what's wrong with the current approaches, and if there are other alternatives that would make them easier or more palatable.

**Comment by epage** [comment]
Yeah, custom arguments on `cargo test` is very tricky.  Custom arguments with custom test harness (`cargo test -- --custom`) would be viable.  I am exploring my own general purpose custom test harness with a pluggable CLI (assert-rs/libtest2#159).  Note that you would most likely want the same set of arguments on all test binaries so you don't have to select specific ones as all of the others would error.

I'm going to go ahead and close this in favor of this being done in custom test harnesses.  if there is a reason for us to re-evaluate, let us know!
